### PR TITLE
Perform hydro checks prior to the simulation [ANT-1720] 

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/application/include/antares/application/ScenarioBuilderOwner.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+#include "antares/study/study.h"
+
+namespace Antares {
```
Comment (payetvin) [inline src/solver/application/include/antares/application/ScenarioBuilderOwner.h]: namespace Antares::Solver

File: src/solver/hydro/management/HydroInputsChecker.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        //            prepareEffectiveDemand();
+        //            prepareMonthlyOptimalGenerations(randomReservoirLevel,
+        //            year);
+        //-----------------------------
+    }
+}
+
+bool HydroInputsChecker::checksOnGenerationPowerBounds(uint year) const
+{
+    return (checkMinGeneration(year) && checkGenerationPowerConsistency(year)) ? true : false;
```
Comment (flomnes) [inline src/solver/hydro/management/HydroInputsChecker.cpp]:
```suggestion
    return checkMinGeneration(year) && checkGenerationPowerConsistency(year);
```

File: src/solver/hydro/management/daily.cpp [hunk 3]
```diff
@@ -227,7 +227,7 @@ inline void HydroManagement::prepareDailyOptimalGenerations(
 {
     const auto srcinflows = area.hydro.series->storage.getColumn(y);
 
-    auto& data = tmpDataByArea_[&area];
+    auto& data = area.hydro.hydro_management_data[y];
```
Comment (flomnes) [inline src/solver/hydro/management/daily.cpp]: This is actually a pretty bad idea, since `area` is shared by threads you'll get race conditions, etc.

**Review by flomnes — CHANGES_REQUESTED** [review]
I think (and @guilpier-code & @sylvlecl will agree) that adding temporary data to a `Study` is a bad idea. It would be best to keep it outside as it is now.

**Comment by a-zakir** [comment]
> I think (and @guilpier-code & @sylvlecl will agree) that adding temporary data to a `Study` is a bad idea. It would be best to keep it outside as it is now.

actually tmpDataByArea is not temporary

File: src/solver/hydro/management/daily.cpp (hunk 3 above)
Comment (a-zakir) [reply src/solver/hydro/management/daily.cpp]: ok

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx [hunk 4]
```diff
@@ -315,7 +317,7 @@ void ISimulation<ImplementationType>::run()
 
     ImplementationType::setNbPerformedYearsInParallel(pNbMaxPerformedYearsInParallel);
 
-    TSGenerator::ResizeGeneratedTimeSeries(study.areas, study.parameters);
+    // TSGenerator::ResizeGeneratedTimeSeries(study.areas, study.parameters);
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/solver.hxx]: remove commented code

**Review by flomnes — COMMENTED** [review]
1. `hydro_management_data` -> `managementData`

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx (hunk 4 above)
Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx]: done

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 5]
```diff
@@ -28,6 +28,50 @@
...
+    double MLE[12];
+    //! Monthly optimal generation
+    double MOG[12];
+    //! Monthly optimal level
+    double MOL[12];
+    //! Monthly target generations
+    double MTG[12];
+    //! inflows
+    double inflows[12];
+    //! monthly minimal generation
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: Use `std::array` instead ?

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 6]
```diff
@@ -28,6 +28,50 @@
...
+    std::array<double, 12> mingens;
+
+    //! Net demand, for each day of the year, for each area
+    double DLN[dayYearCount];
+    //! Daily local effective load
+    double DLE[dayYearCount];
+    //! Daily optimized Generation
+    double DOG[dayYearCount];
+    //! daily minimal generation
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: `std::array` ?

File: src/solver/application/ScenarioBuilderOwner.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+
+#include "antares/application/ScenarioBuilderOwner.h"
+#include <antares/antares/fatal-error.h>
+#include "antares/solver/simulation/apply-scenario.h"
+#include "antares/solver/simulation/timeseries-numbers.h"
+#include "antares/solver/ts-generator/generator.h"
```
Comment (flomnes) [inline src/solver/application/ScenarioBuilderOwner.cpp]: Use `< >`

File: src/solver/application/application.cpp [hunk 8]
```diff
@@ -33,6 +33,7 @@
 #include <antares/sys/policy.h>
 #include <antares/writer/writer_factory.h>
 #include "antares/antares/version.h"
+#include "antares/application/ScenarioBuilderOwner.h"
 #include "antares/config/config.h"
 #include "antares/file-tree-study-loader/FileTreeStudyLoader.h"
 #include "antares/signal-handling/public.h"
```
Comment (flomnes) [inline src/solver/application/application.cpp]: Use `< >`

File: src/solver/application/include/antares/application/ScenarioBuilderOwner.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+#include "antares/study/study.h"
```
Comment (flomnes) [inline src/solver/application/include/antares/application/ScenarioBuilderOwner.h]: Forward declaration is enough here (only refs to `Data::Study`, no need to know what's inside)

File: src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+#include <antares/study/area/area.h>
```
Comment (flomnes) [inline src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h:22]: You may include study.h here

File: src/solver/hydro/management/HydroInputsChecker.cpp [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+        //---------------------  May be later ? //
+        //            prepareNetDemand(year,
+        //            parameters_.mode,
+        //            scratchmap);
+        //            prepareEffectiveDemand();
+        //            prepareMonthlyOptimalGenerations(randomReservoirLevel,
+        //            year);
+        //-----------------------------
```
Comment (flomnes) [inline src/solver/hydro/management/HydroInputsChecker.cpp]: Remove this commented code ?

File: src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+{
+
+class HydroInputsChecker
+{
+public:
+    HydroInputsChecker(Antares::Data::Study& study, Solver::IResultWriter& resultWriter);
+    void Execute();
+
+private:
+    Antares::Data::Study& study_;
```
Comment (flomnes) [inline src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h]: I see that you already have a few references to member data of the Study class. If you can, avoid referencing Study, since it's a "god class", it makes it difficult to see what is actually meant to be used.

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 13]
```diff
@@ -28,6 +28,63 @@
...
+    //! Monthly optimal level
+    std::array<double, 12> MOL;
+    //! Monthly target generations
+    std::array<double, 12> MTG;
+};
+
+// TODO Rename
+struct HydroSpecific
+{
+    HydroDaily daily;
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: You may want to use `double`s in `HydroDaily`, and an `std::array<HydroDaily, 12>` here, so you have 1 array instead of multiple. The same goes for monthly data.

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 14]
```diff
@@ -28,6 +28,63 @@
...
+    //! Net demand, for each day of the year, for each area
+    std::array<double, dayYearCount> DLN;
+    //! Daily local effective load
+    std::array<double, dayYearCount> DLE;
+};
+
+struct HydroMonthly
+{
+    //! Monthly local effective demand
+    std::array<double, 12> MLE;
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: You may use the comment above to rename these variables. Quite unexpressive right now

File: src/solver/application/include/antares/application/ScenarioBuilderOwner.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+namespace Antares
```
Comment (flomnes) [inline src/solver/application/include/antares/application/ScenarioBuilderOwner.h:23]: Use namespace nesting from cpp17

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 16]
```diff
@@ -28,6 +28,63 @@
 
 namespace Antares::Data
 {
+
+enum dayYearCount
+{
+    //! The maximum number of days in a year
+    dayYearCount = 366
```
Comment (flomnes) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: Use `constexpr` if possible

**Review by flomnes — COMMENTED** [review]
Mostly style & improvement suggestions.

If a function only uses `HydroSpecific::monthly`, you can just pass the `monthly` object. Same for `daily`.

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h (hunk 16 above)
Comment (a-zakir) [reply src/libs/antares/study/include/antares/study/parts/hydro/container.h]: thank you

**Comment by a-zakir** [comment]
> Mostly style & improvement suggestions.
> 
> If a function only uses `HydroSpecific::monthly`, you can just pass the `monthly` object. Same for `daily`.

data structs have changed, i think the code is tidy enough & Antares::Data::TimeDependantHydroManagementData is already compact

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 17]
```diff
@@ -28,6 +28,60 @@
 
 namespace Antares::Data
 {
+
+    //! The maximum number of days in a year
+constexpr size_t dayYearCount = 366;
+
+struct HydroDailyLoadDemand
```
Comment (guilpier-code) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: **HydroDailyLoadDemand** : there is some kind of repetition here.
**load** and **demand** are synonymous.
**HydroDailyDemand** is enough. 
Given the context we're in (hydro), **DailyDemand** could even do the job.

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 18]
```diff
@@ -28,6 +28,60 @@
...
+{
+    //! Net demand, for each day of the year, for each area
+    double DLN = 0.;
+    //! Daily local effective load
+    double DLE = 0.;
+};
+
+struct HydroMonthlyDemandGeneration
```
Comment (guilpier-code) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h]: **HydroMonthlyDemandGeneration** : 
- This **struct** gather things that are quite different : **MLE** is a demand (or load if you prefer), whereas **MTG** is an hydro generation, and **MOL** is an hydro level.
- The name of the **struct** doesn't make sense to me. What is a hydro monthly demand generation ?

In the end, what we want to compute here is a hydro monthly generation target.
So a **struct** named **MonthlyGenerationTargetData** would make more sense.

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx [hunk 19]
```diff
@@ -1015,65 +989,74 @@ void ISimulation<ImplementationType>::loopThroughYears(uint firstYear,
...
     pQueueService->maximumThreadCount(pNbMaxPerformedYearsInParallel);
+    HydroInputsChecker hydroInputsChecker(study);
 
-    // Loop over sets of parallel years
-    std::vector<setOfParallelYears>::iterator set_it;
-    for (set_it = setsOfParallelYears.begin(); set_it != setsOfParallelYears.end(); ++set_it)
+    // Loop over sets of parallel years to check hydro inputs
+    for (const auto& batch: setsOfParallelYears)
+    {
+        // if refresh break
```
Comment (guilpier-code) [inline src/solver/simulation/include/antares/solver/simulation/solver.hxx]: About `// if refresh break` : 
This comment is less clear than the code. 
Code is clear enough. We should remove the comment.

File: src/solver/application/ScenarioBuilderOwner.cpp [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+*/
+
+#include <antares/antares/fatal-error.h>
+#include <antares/application/ScenarioBuilderOwner.h>
+#include "antares/solver/simulation/apply-scenario.h"
+#include "antares/solver/simulation/timeseries-numbers.h"
+#include "antares/solver/ts-generator/generator.h"
+#include "antares/study/study.h"
+
+Antares::Solver::ScenarioBuilderOwner::ScenarioBuilderOwner(Data::Study& study):
```
Comment (guilpier-code) [inline src/solver/application/ScenarioBuilderOwner.cpp:29]: Class **ScenarioBuilderOwner** : not sure we need a class in order to call free functions about TS numbers.
Is it not a bit over engineered ?

File: src/solver/application/ScenarioBuilderOwner.cpp [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    study_(study)
+{
+}
+
+void Antares::Solver::ScenarioBuilderOwner::callScenarioBuilder()  {
```
Comment (guilpier-code) [inline src/solver/application/ScenarioBuilderOwner.cpp:34]: Class **ScenarioBuilderOwner** : 
This name suggests the class is about **scenario builder** and only **scenario builder**.
Looking closely, this class is about : 
- **resizing** time series,  
- **resizing** TS numbers to the number of MC years, 
- **generating** TS numbers, 
- and finally applying (taking only TS numbers chosen by user) the scenario builder rules.

The 3 first bullets are not really related to scenario builder. 
It is misleading.
Either we think harder to find a better name (I'm afraid there isn't any), or we leave the original function calls.

File: src/solver/hydro/management/HydroInputsChecker.cpp [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+    {
+        checked_years_[year] = false;
+    }
+}
+
+void HydroInputsChecker::Execute(uint year)
+{
+    if (!checked_years_[year])
+    {
+        checked_years_[year] = true;
```
Comment (guilpier-code) [inline src/solver/hydro/management/HydroInputsChecker.cpp]: **checked_years_** : seems useless.
Why do we need this data member ?

If it turns out to be useless, remove data members **firstYear_** and **endYear_**, useless too.

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx [hunk 23]
```diff
@@ -1015,65 +989,74 @@ void ISimulation<ImplementationType>::loopThroughYears(uint firstYear,
 
     // Number of threads to perform the jobs waiting in the queue
     pQueueService->maximumThreadCount(pNbMaxPerformedYearsInParallel);
+    HydroInputsChecker hydroInputsChecker(study);
 
-    // Loop over sets of parallel years
-    std::vector<setOfParallelYears>::iterator set_it;
-    for (set_it = setsOfParallelYears.begin(); set_it != setsOfParallelYears.end(); ++set_it)
+    // Loop over sets of parallel years to check hydro inputs
+    for (const auto& batch: setsOfParallelYears)
```
Comment (guilpier-code) [inline src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: About block : 
```C++
for (const auto& batch: setsOfParallelYears)
{
    if (batch.regenerateTS)
    {
        break;
    }
    for (auto year: batch.yearsIndices)
    {
        hydroInputsChecker.Execute(year);
    }
}
```
Is this block useful ?
Indeed, there is a further loop over batches, and there's a check on hydro input as well.
With this first block, don't we double check ?

Comment (guilpier-code) [inline src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: **I'm sure you're aware** that, by checking hydro input just before a batch of MC years is run, we still have the problem for which we opened this PR :  if a year's hydro data is not valid inside a batch, we stop in the middle of a simulation. 
And if the simulation is run in sequential mode, we gain nothing.

I know that **hydro TS refresh** is the cause of this poor gain we've here : it prevents us to move checks on hydro input before any batch of MC year is run.

Unless TS refresh disappears, we'll be hampered with this project.

Possible solutions : 
- As discussed with @a-zakir, a temporary solution could be : 
  - to reset the hydro random generator after TS were generated a first time **only for needs of hydro input checks**, 
  - and then to generate the hydro Ts a second time for real hydro management.
- If we succeed in having the checks on hydro input before running the MC years batch, we still have to store hydro data used for each area and year. Why not keep the hydro checks inside MC years, but just not run years for which data is not valid ?

File: src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+private:
+    Data::AreaList& areas_;
+    const Data::Parameters& parameters_;
+    const Date::Calendar& calendar_;
+    Data::SimulationMode simulationMode_;
+    const uint firstYear_;
+    const uint endYear_;
+    PrepareInflows prepareInflows_;
+    MinGenerationScaling minGenerationScaling_;
+    bool refresh_ = false;
```
Comment (guilpier-code) [inline src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h]: - `bool refresh_ = false;`
- `const Data::Parameters& parameters_`
- `Data::SimulationMode simulationMode_;`

These data members are unused, and should be removed

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h [hunk 25]
```diff
@@ -28,6 +28,60 @@
...
+
+//!  Hydro Management Data for a given area
+struct TimeDependantHydroManagementData
+{
+    std::array<HydroDailyLoadDemand, dayYearCount> daily{0};
+    std::array<HydroMonthlyDemandGeneration, 12> monthly{0};
+};
+
+//! Area Hydro Management Data for a given year
+struct AreaDependantHydroManagementData
```
Comment (guilpier-code) [inline src/libs/antares/study/include/antares/study/parts/hydro/container.h:64]: **struct TimeDependantHydroManagementData** : 
This **struct** is used only during hydro ventilation (hydro management if you prefer).
It's not used when checking hydro data.
So we could ask : why is it defined here, and not in class **HydroManagement** ?

Same for other struct above.

**Review by guilpier-code — DISMISSED** [review]

File: src/solver/application/ScenarioBuilderOwner.cpp (hunk 20 above)
Comment (a-zakir) [reply src/solver/application/ScenarioBuilderOwner.cpp:29]: i like classes, it's chic

File: src/solver/application/ScenarioBuilderOwner.cpp (hunk 21 above)
Comment (a-zakir) [reply src/solver/application/ScenarioBuilderOwner.cpp:34]: conservative are you?

File: src/solver/hydro/management/HydroInputsChecker.cpp (hunk 22 above)
Comment (a-zakir) [reply src/solver/hydro/management/HydroInputsChecker.cpp]: i'm the batch n°2 (with TS refresh 💧 ) , the batch before me didn't refresh so he was checked in the 1st loop right?
[myself batch n2] i'm thrown off the 1st loop and my checks are done in the 2nd loop

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx (hunk 23 above)
Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: nope thanks to checked_years_

Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: i understand it's a bit confusing

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h (hunk 18 above)
Comment (a-zakir) [reply src/libs/antares/study/include/antares/study/parts/hydro/container.h]: renamed

File: src/libs/antares/study/include/antares/study/parts/hydro/container.h (hunk 17 above)
Comment (a-zakir) [reply src/libs/antares/study/include/antares/study/parts/hydro/container.h]: renamed

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx (hunk 23 above)
Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: Ts generation on the fly and refresh will be abandoned soon, therefore the outer loop ensure that hydro checks are done before the simulation loop

File: src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h (hunk 24 above)
Comment (a-zakir) [reply src/solver/hydro/include/antares/solver/hydro/management/HydroInputsChecker.h]: thanks they were used in a previous version

File: src/solver/application/ScenarioBuilderOwner.cpp (hunk 21 above)
Comment (guilpier-code) [reply src/solver/application/ScenarioBuilderOwner.cpp:34]: > conservative are you?

I'm sorry, I don't understand your answer, could be more specific ?

File: src/solver/hydro/management/HydroInputsChecker.cpp (hunk 22 above)
Comment (flomnes) [reply src/solver/hydro/management/HydroInputsChecker.cpp]: Yes you can just ignore refresh

File: src/solver/application/ScenarioBuilderOwner.cpp (hunk 20 above)
Comment (guilpier-code) [reply src/solver/application/ScenarioBuilderOwner.cpp:29]: I've nothing against classes, I'm even a fan of classes. 
But this class contain heterogeneous things (see the [comment below](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2132#discussion_r1646523749) ), and its name does

File: src/solver/simulation/include/antares/solver/simulation/solver.hxx (hunk 23 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: Still on this hydro checking **loop over batches** : 
TS in general, and hydro TS in particular are generated by a call to **regenerateTimeSeries**.
When running this loop (before main loop over batches), **hydro TS haven't been generated** yet, they were only resized by calling **TSGenerator :: ResizeGeneratedTimeSeries**.
So the checks on hydro input are based on hydro TS for which every value is **zero**.

Comment (a-zakir) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: not all time regenerateTimeSeries is called, it's the case when (generate & refresh) are used. This options will be obsolete soon and all TS won"t be zero

Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/solver.hxx:968]: Ok, my bad, I missed something

File: src/solver/hydro/management/HydroInputsChecker.cpp (hunk 22 above)
Comment (flomnes) [reply src/solver/hydro/management/HydroInputsChecker.cpp]: checked_years_ was removed.

**Review by flomnes — APPROVED** [review]
